### PR TITLE
[llama] Update kv cache to have read/write functions

### PR DESCRIPTION
**Pull request by rsuderman**
Made the interfaces of both caches line up. This allows us to interface with the caches via their utility functions instead of modifying the model behavior. Some roughness still exists in their parameters but the irrelevant details are ignored for each implementation.

Need to still add some slicing / ignoring on the page ids to make more flexible.

### CODE REVIEW

File: sharktank/sharktank/layers/kv_cache.py [hunk 1]
```diff
@@ -113,15 +115,109 @@ def allocate(self, *, bs: int) -> list[torch.Tensor]:
...
+        return [
+            ops.reshard_split(allocation, dim=2, count=self.shard_count)
+            for allocation in allocations
+        ]
+
+    def read(
+        self,
+        state: list[Union[torch.Tensor, SplitPrimitiveTensor]],
+        *,
+        read_into_partitions: list[Union[torch.Tensor, SplitPrimitiveTensor]],
```
Comment (sogartar) [inline sharktank/sharktank/layers/kv_cache.py:144]: Why are we not reading into this?

**Review by sogartar — CHANGES_REQUESTED** [review]

**Comment by dan-garvey** [comment]
We needed this so badly, thanks Rob

File: sharktank/sharktank/layers/kv_cache.py (hunk 1 above)
Comment (rsuderman) [reply sharktank/sharktank/layers/kv_cache.py:144]: There is no value in placing in a destination block. Ultimately this should fuse with the attention block (this is the premise of paged attention fusion)

**Review by dan-garvey — APPROVED** [review]
LGTM

**Review by sogartar — APPROVED** [review]

**Pull request merged**